src/bartiq/symbolics/ast_parser.py

```diff
@@ -0,0 +1,302 @@
+        designator.
+        """
+        if isinstance(node.func, ast.Name) and node.func.id == "Port":
+            return self.interpreter.create_parameter((f"#{_resolve_value(node.args[0])}",))
+        if isinstance(node.func, ast.Attribute) and node.func.attr == "Port":
+            return self.interpreter.create_parameter(
+                (f"{_resolve_value(node.func.value)}.#{_resolve_value(node.args[0])}",)
+            )
+        else:
+            return self.interpreter.create_function(
+                (_resolve_value(node.func), list(map(self.convert_node, node.args)))
+            )
+
+    @convert_node.register(ast.Name)
+    @convert_node.register(ast.Attribute)
+    def _(self, node):
+        """Variant of convert_node for ast.Name and ast.Attribute, which are nodes representing parameter."""
+        return self.interpreter.create_parameter((_resolve_value(node),))
+
+
+@singledispatch
+def _resolve_value(value_node):
+    """Given a node, resolve it to a namespaced name.
+
+    This function works for ast.Name nodes, as well as (possibly nested) ast.Attribute nodes, for
+    which it recurses into their value part.
+    Also, if an ast.Call node is a part of attribute lookup, this is an error, unless the function
+    being called is a wildcard - in which case we simply use ~ symbol in the namespaced symbol name.
+    """
+    raise NotImplementedError(f"Unexpected node found when resolving attribute lookup: {type(value_node)}")
+
+
+@_resolve_value.register
+def _(value_node: ast.Name):
+    return _restore_name(value_node.id)
+
+
+@_resolve_value.register
+def _(value_node: ast.Attribute):
+    return f"{_resolve_value(value_node.value)}.{_restore_name(value_node.attr)}"
+
+
+@_resolve_value.register
+def _(value_node: ast.Call):
+    # Assertions here are to silence Mypy and describe the expected input.
+    # We should never get any assertion errors here, unless the expression doesn't match our grammar.
+    assert isinstance(value_node.func, ast.Name)
+    if value_node.func.id != "wildcard":
+        raise ValueError("Should never encounter function call other than wildcard() in the attribute lookup")
+    assert isinstance(value_node.args[0], ast.Name)
+    return f"{value_node.args[0].id}~" if value_node.args else "~"
+
+
+def parse(expression: str, interpreter: Interpreter):
+    """Parse given mathematical using given interpreter to create expression primitives.
+
+    Args:
+        expression: An expression to parse.
+        interpreter: An interpreter providing methods for constructing numbers, symbols, and functions.
+
+    Returns:
+        A result of interpreting the whole expression (which depends on what the interpreter produces).
+    """
+    preprocessed_expression = _preprocess(expression)
+    return NodeConverter(interpreter).convert_node(ast.parse(preprocessed_expression))
+
+
+def parse_to_sympy(expression: str, debug=False):
```

Comment:
[minor] lacks output type

Reply:
Also, TBH, wouldn't this function fit the `sympy_backends.py` file better? Then we don't need sympy in this file at all and have clearer separation?

Reply:
Fixed!